### PR TITLE
feat(calendar): add one-click Add to Calendar button to event cards

### LINKED ISSUE #19: [FEATURE] : Add to Calendar (ICS / Google Calendar / Outlook) for Each Event
### Description

Currently, users browsing events have no way to save them directly to their personal calendar. They must manually copy the event title, date, time, and location into Google Calendar, Apple Calendar, or Outlook — which is unnecessary friction and likely reduces event attendance and engagement.

This feature adds a one-click "Add to Calendar" button on each event card that:

- Downloads a standard `.ics` file (compatible with Google Calendar, Apple Calendar, and Outlook)
- Works entirely client-side via `Blob + URL.createObjectURL`, so it is fully compatible with static hosting like GitHub Pages with no backend required
- Optionally provides direct deep-link buttons for Google Calendar and Outlook Web for users who prefer not to download a file

 ---

### Use Case
- A visitor sees an interesting event and wants to save it without manually entering details
- A mobile user wants to tap "Add to Calendar" and have it open directly in their phone's calendar app
- An organizer shares the events page and wants maximum conversion from "interested" to "attending"
- A user wants to bulk-download all upcoming events as a single `.ics` file to import at once

 ---

### Proposed Implementation
 1. ICS Generator Utility — `src/utils/icsGenerator.js`
- Pure functions: `parseTime(), formatICSDate(), escapeICSText(), generateICS()`
- `downloadICS()` as the only DOM-touching function `(triggers Blob download)`
- Supports 12-hour and 24-hour time formats

 2. EventCard Update — `src/components/EventCard.jsx`
- Add an `"Add to Calendar"` `<button>` inside a new `.event-card__actions` wrapper
- Calls `downloadICS(event)` on click
- Existing "Learn more" link moves into the same actions row

Tests — `src/utils/icsGenerator.test.js`
1. parseTime handles 12hr AM/PM and 24hr formats
2.  parseTime handles edge cases (midnight, noon)
3. escapeICSText escapes commas, semicolons, backslashes, newlines
4. generateICS contains correct DTSTART / DTEND
5. generateICS respects custom duration
6. generateICS uses CRLF line endings
7. generateICS escapes special characters in SUMMARY

### PR DESCRIPTION
**Pull request by nidhipawar5**
## [FEATURE] : Add to Calendar (ICS / Google Calendar / Outlook) for Each Event #19

Resolves #19

Users had no way to save events directly to their calendar from the event board, they had to manually copy the title, date, time, and location into their calendar app. This PR adds a one-click "Add to Calendar" button to each event card.

Clicking the button downloads a `.ics` file that works with Google Calendar, Apple Calendar, and Outlook. The whole thing runs client-side using `Blob + URL.createObjectURL`, so there's no backend needed and it works fine on GitHub Pages.

Changes made:
- Added `src/utils/icsGenerator.js` with pure utility functions: `parseTime`, `formatICSDate`, `escapeICSText`, `generateICS`, and `downloadICS`
- Updated `EventCard.jsx` to include the button in a new `.event-card__actions` row, with the existing "Learn more" link sitting alongside it
- Added styles for the actions row and button in `index.css`
- Added `src/utils/icsGenerator.test.js` with 24 unit tests

## How to test these changes

- Run `npm install` then `npm run generate` then `npm run dev`
- Open `localhost:5173/du-event-board/` in your browser
- You should see an "Add to Calendar" button on each event card next to the "Learn more" link
- Click it,  a `.ics` file should download named after the event
- Import the file into Google Calendar, Apple Calendar, or Outlook and verify the event title, date, time, and location appear correctly
- Run `npm test` to confirm all 39 tests pass

## Pull Request checklists

This PR is a:
- [ ] bug-fix
- [x] new feature
- [ ] maintenance

About this PR:
- [x] it includes tests.
- [x] the tests are executed on CI.
- [x] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:
- [x] I have reviewed the changes and it contains no misspelling.
- [x] The code is well commented, especially in the parts that contain more complexity.
- [x] New and old tests passed locally.

## Additional information

The `.ics` generator follows RFC 5545 — it uses CRLF line endings, escapes commas, semicolons, and backslashes in text fields, and sets a stable UID per event so re-importing the same file doesn't create duplicate calendar entries.

## Summary
Closes #19

Adds a one-click **Add to Calendar** button to every event card that downloads a standard `.ics` file, compatible with Google Calendar, Apple Calendar, and Outlook. Works entirely client-side via `Blob + URL.createObjectURL`, so it is fully compatible with static GitHub Pages hosting.

## Changes
- `src/utils/icsGenerator.js` - new utility module (pure functions only;
  `downloadICS` is the single DOM-touching function)
- `src/utils/icsGenerator.test.js`- 24 unit tests covering all cases
  specified in the issue
- `src/components/EventCard.jsx` - Add to Calendar button + actions row;
  existing "Learn more" link moved into the same row
- `src/index.css` - styles for `.event-card__actions` and
  `.event-card__calendar-btn`

## Testing
- `npm test` → 39/39 tests pass (24 new + 15 existing)

Screenshot 2026-03-21 at 11 57 13 PM

- `npm run lint` → 0 errors
- Manually tested .ics download in Chrome and Firefox; imported into
  Google Calendar successfully

## Screenshots

Screenshot 2026-03-21 at 11 58 23 PM

## Please note:

The optional Google Calendar and Outlook Web deep-link buttons mentioned in the issue are not included in this PR and can be addressed as a follow-up.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1694,6 +1696,7 @@
       "integrity": "sha512-z9VXpC7MWrhfWipitjNdgCauoMLRdIILQsAEV+ZesIzBq/oUlxk0m3ApZuMFCXdnS4U7KrI+l3WRUEGQ8K1QKw==",
       "dev": true,
       "license": "MIT",
+      "peer": true,
```
Comment (yuvimittal) [inline package-lock.json:1699]: you have added these changes?

**Comment by yuvimittal** [comment]
@nidhipawar5 , i have rebased the branch.. but there are few changes where i have added comments, can you please look into it ?

File: data/.geocode_cache.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "Porto Alegre": [-30.0324999, -51.2303767],
```
Comment (yuvimittal) [inline data/.geocode_cache.json:2]: how are we adding the location coordinates?
